src/oscar/utils/srcsets.py

```diff
@@ -0,0 +1,114 @@
+        # uploaded source image.
+        if not do_upscale and width >= source.width:
+            return
+
+        # give the options of passing in a custom image processor.
+        image_processor = image_processor or get_thumbnailer().generate_thumbnail
+
+        return image_processor(source, **dict({'size':calc_img_dim(source, width)},**options))
+def get_srcset(source, image_type=None, image_processor=None, do_upscale=False,
+        settings=settings, **options):
+        gen_sizes = _get_srcset_sizes(get_settings(), image_type=image_type)
+        gen_create_image = (
+            (image_type, size, width, get_srcset_image(
+                source, width, 
+                image_type=image_type, 
+                image_processor=image_processor, 
+                do_upscale=do_upscale,
+                settings=settings, 
+                **options
+            )) for image_type, size, width in gen_sizes)
+        return (
+            SrcsetImage(image, size, width, image_type) 
+            for image_type, size, width, image in gen_create_image
+            if image is not None
+        )
+
+class SrcsetImage():
+    
+    def __init__(self, image, size, width, image_type):
+        self.image = image
+        self.size = size
+        self.width = width
+        self.image_type = image_type
+        return
+    def __str__(self):
+        return f'{image.url} {width}w'
+
+class SrcsetCollection():
+    
+    def __init__(self, instance, srcset_images):
+        self.instance = instance
+        self._srcset = srcset_images
+        return
+    def __getattr__(self, name):
+        rtn = next((si for si in self._srcset if si.size==name), None) \
+            or [si for si in self._srcset if si.image_type==name]
+        if rtn: 
+            return rtn
+        raise AttributeError()
+    def __getitem__(self,name):
+        return next((si for si in self._srcset if si.width==name), None) \
+            or self.__getattr__(name)
+    def __str__(self):
+        return str(self._srcset)
+    def __len__(self):
+        return len(self._srcset)
+
+class SrcSetDescriptor:
+
+    def __init__(self, source_field, image_type=None, image_processor=None, do_upscale=False,
+        settings=settings, **options):
+        self.source_field = source_field
+        self.image_type = image_type
+        self.image_processor = image_processor or get_thumbnailer().generate_thumbnail
+        self.do_upscale=do_upscale
+        self.settings = settings or get_settings()
+        self.options = options
+        return
+
+    def __get__(self, instance, cls=None):
+        
+        srcset = get_srcset(
+            getattr(instance, self.source_field), 
+                image_type=self.image_type, 
+                image_processor=self.image_processor, 
+                do_upscale=self.do_upscale,
+                settings=self.settings, 
+                **self.options
+        )
+        return SrcsetCollection(instance, list(srcset))
```

Comment:
I don't know why you made get_srcset return a generator to then immediately cast it to a list. You might want to use  list comprehension. I do think an implementation of srcset should be lazy and delay creating any images until the image url is actually accessed, which is not what happens here. Correct me if I'm wrong.